### ISSUE TITLE
[bug] Loading a video/audio as an asset does not work

### ISSUE THREAD

**Issue by firstdorsal** [tauri-apps/tauri#3725]
### Describe the bug

Loading a video as an asset does not work.

I followed the guide here: https://tauri.studio/docs/api/js/modules/tauri

It works fine with image files (png/jpg etc.) but not with videos. 
I tried loading the video https://www.w3schools.com/html/mov_bbb.mp4 directly from the URL, and this worked without problems.
But using the `asset:///home/paul/Downloads/mov_bbb.mp4` URL this does not work
Loading images from the same folder does work.

In the network panel you can see that the request failed but no further reason is specified
![image](https://user-images.githubusercontent.com/10563452/158854074-2a3b165e-8490-4ae2-b418-2ff682bac0a1.png)

![image](https://user-images.githubusercontent.com/10563452/158854189-c1b2de0a-6fd5-4b1d-9bb3-a0a8bc8ccf0b.png)

The loading of assets this way is IMO not very good explained and when it fails, for some reason, it is very difficult to find out why.

Thank you for creating Tauri, it is awesome!

### Platform and versions

```shell
Operating System - Arch Linux, version Rolling Release X64

Node.js environment
  Node.js - 17.7.1
  @tauri-apps/cli - 1.0.0-rc.6
  @tauri-apps/api - 1.0.0-rc.2

Global packages
  npm - 8.5.4
  pnpm - Not installed
  yarn - 1.22.17

Rust environment
  rustup - 1.24.3
  rustc - 1.59.0
  cargo - 1.59.0
  toolchain - stable-x86_64-unknown-linux-gnu 

App directory structure
/dist
/node_modules
/src-tauri
/src
/.git
/public

App
  tauri - 1.0.0-rc.4
  tauri-build - 1.0.0-rc.4
  tao - 0.6.4
  wry - 0.13.3
  build-type - build
  CSP - default-src tauri: asset: https://asset.localhost
  distDir - ../dist
  devPath - http://localhost:3000/
  framework - React
```

**Comment by red010182** [comment]
This config works for me.
```
    "allowlist": {
      ...
      "protocol": {
        "asset": true,
        "assetScope": ["**"]
      }
    },
    "security": {
      "csp": "default-src 'self'; img-src 'self' asset: https://asset.localhost; media-src 'self' asset: https://asset.localhost"
    }

```

---
Edit:

Don't do this. It's dangerous: https://github.com/tauri-apps/tauri/issues/3735#issuecomment-1072978250

**Comment by FabianLars** [comment]
As i said in the other issue, you shouldn't configure the scope like this (Especially since we are working on a fix).
You're better of by doing it more like this:
1) Configure it correctly and as strict as possible.
2) Use dialogs and file-drop events to dynamically extend it.
3) Use rust to manually extend the scope at runtime as a last resort.

About scope errors: It should be logged to the terminal (in `tauri dev`) and we return a 403 response (with an empty body) on scope related errors

**Comment by firstdorsal** [comment]
This is not a scope issue. My scope already looks like this:
```json
 "protocol": {
      "all": false,
      "asset": true,
      "assetScope": ["*"]
},
```
For my project this is necessary as the file explorer that i'm building requires the capability to, you know, display files :p. 
Images do work via asset://, text like files and pdfs are loaded through rust but videos won't load with the assets protocol.

**Comment by FabianLars** [comment]
Ah whoops, didn't see that you were 2 different people, sorry.

**Comment by red010182** [comment]
Did you try adding media-src in csp? Add `media-src 'self' asset: https://asset.localhost` works for me.

**Comment by firstdorsal** [comment]
@red010182
The `default-src` already does allow assets: `default-src tauri: asset: https://asset.localhost`
I tried it anyway but it is not working. Also loading assets from the web wouldn't work if that was the problem.

**Comment by tedbyron** [comment]
@firstdorsal do you have a reproduction? I have a video conversion app I'm making and I'm able to display user-uploaded videos (from tauri dialog) in a `<video>` element with the following config:

```jsonc
{
  // ...
  "protocol": {
    "asset": true
  },
  // ...
  "security": {
    "csp": "default-src: 'self'; media-src 'self' asset:"
  }
}
```

**Comment by lucasfernog** [comment]
I can reproduce this issue, seems to be a webkit2gtk problem.

**Comment by dzervas** [comment]
Is there a workaround that we can use? Or we just wait for a proper fix?

**Comment by firstdorsal** [comment]
@tedbyron 
> from dialog

If you mean using the "Native" browser dialog to select a video then this is not related I think.

@lucasfernog great, good to know! Is this working on Win/Mac?

**Comment by lucasfernog** [comment]
I never had issues on Windows and macOS. And I don't know about a workaround :/

**Comment by lucasfernog** [comment]
@wusyong was investigating this issue, I think it's a problem with latest webkit2gtk.

**Comment by joehillen** [comment]
Is `https://asset.localhost/<filepath>` supposed to work also? I get "Connection refused" with that but nothing at all with `asset://`.

**Comment by joehillen** [comment]
Also, is there an upstream bug report? I can't find one.

**Comment by FabianLars** [comment]
> Is `https://asset.localhost/<filepath>` supposed to work also? I get "Connection refused" with that but nothing at all with `asset://`.

`asset://` is Linux and macOS and `https://asset.localhost/` is Windows.

**Comment by keiya01** [comment]
I tried to open [this page](https://yari-demos.prod.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/video/_sample_.single_source.html) on wry like the following.

```rs
let webview = WebViewBuilder::new(window)?.with_url("https://yari-demos.prod.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/video/_sample_.single_source.html")?;
```

But this shows the empty video screen .

https://user-images.githubusercontent.com/34934510/181906813-6ad3646c-f89a-42d0-9477-e35686b19435.mov

edit: oops, this is my mistakes. webkitgtk seems to need to [install gstreamer](https://gstreamer.freedesktop.org/documentation/installing/on-linux.html?gi-language=c)...

**Comment by keiya01** [comment]
Maybe issue is here for webkitgtk
https://bugs.webkit.org/show_bug.cgi?id=146351

**Comment by lucasfernog** [comment]
It's weird 'cause I was sure it worked on previous webkit2gtk releases :(

**Comment by Robinerra** [comment]
Loading audio with the asset protocol is also broken

**Comment by wusyong** [comment]
It seems fixed in 2.36.4.
https://webkitgtk.org/2022/07/28/webkitgtk2.36.5-released.html
I plan to add headers support again when I have some time. Will test this too.

**Comment by loicortola** [comment]
Hi all,
I don't know if it helps, I have just tested a basic video sample using the asset protocol. With a regular up-to-date ubuntu release, using webkitgtk2.36.6.
Videos are still not playing on my station. Hopefully the next webkitgtk releases will have it fixed :)

**Comment by trippjoe** [comment]
Commenting the function name so others can find this issue when they search for `convertFileSrc`.

**Comment by rribou** [comment]
Still no workaround?

**Comment by mrdududu** [comment]
I have same problem but with images on MacOS
This code fix this issue
`"protocol": {
  "asset": true
},
"security": {
  "csp": "default-src: 'self'; media-src 'self' asset:"
},
`

**Comment by Soremwar** [comment]
Same issue, only for videos on Linux

**Comment by iohzrd** [comment]
I'm experiencing this on Arch Linux (gnome) as well.
Video was working in my app a couple days ago, now it's blank and my code hasn't changed. So I believe it's an issue in webkitgtk. 
My system currently has webkit2gtk v2.38.2-2.

**Comment by iohzrd** [comment]
Update:
I had recently reinstalled my OS and was missing some packages.
In my case, this issue was fixed by installing gstreamer and gst-plugins(gst-libav gst-plugins-bad gst-plugins-base gst-plugins-good gst-plugins-ugly).
Which is not mentioned in the getting started / docs.
Might be nice to add mention of this somewhere.

**Comment by wusyong** [comment]
Looks like the issue still remains and I tried my best to write a C example to report back:
https://bugs.webkit.org/show_bug.cgi?id=146351#c5

**Comment by stevenhansel** [comment]
@iohzrd Hey, can you elaborate more on how you fixed it? I upgraded my webkit2gtk to v2.38.2-2 and also installed gstreamer1.0 alongside all of the recommended plugins. It still doesn't want to load the video resource. I'm also getting the local video path using `convertFileSrc`.

**Comment by FabianLars** [comment]
There are 2 issues mixed up here:
1) videos not loading in general, including remote videos (for example from youtube) -> The fix is to install gstreamer
2) local videos not loading via convertFileSrc -> No fix available right now (gstreamer would still be necessary)

**Comment by xxidbr9** [comment]
I got an issue where i need to load image and video inside tauri as an asset where the file is located in randomly path, but i solve that issue by copying the file inside tauri PATH, in my case i use appDataDir and it work just fine

### My example app: 
<img width="1680" alt="Screenshot 2023-02-12 at 21 25 06" src="https://user-images.githubusercontent.com/51733515/218316996-65cce545-711e-48dd-883d-433d44173f0f.png">

I use `react-player` in this example and work fine

### My code 
<img width="1309" alt="Screenshot 2023-02-12 at 21 30 07" src="https://user-images.githubusercontent.com/51733515/218317076-ac6f8e57-fbf7-4492-91f2-a61e4b49e944.png">

**Comment by Nfinished** [comment]
was this removed from the roadmap because there's no intent to support? I'm trying to let users pick a video file but after selection I get `asset protocol not configured to allow the path: Users/nfinished/Desktop/sample.mp4` in console and a 403 in the app.

assetScope is `["*"]` and csp is `default-src 'self'; media-src 'self' asset:`

**Comment by FabianLars** [comment]
Don't mind our project boards yet, we're just playing around with them, trying to find something that works for us. In the latest iteration they only contain PRs which is why this issue was removed.

**Comment by tranxuanthang** [comment]
Came here from #6962, thanks to @FabianLars.

After reading other related issues (this and #2327 and #4496), I thought that this should be fixed in webkit2gtk 2.34.0. I'm facing this issue on webkit2gtk-4.0 2.40. Does this mean newer versions of webkit2gtk break this again? In Windows it still works without issues.

Example snippet (I'll try to create an example repo later):

```javascript
import { Howl } from 'howler'
import { convertFileSrc } from '@tauri-apps/api/tauri'

const filePath = '/home/user/Music/example.mp3'
const assetUrl = convertFileSrc(filePath)
const sound = new Howl({
  src: [assetUrl]
}) // Origin http://localhost:1420 is not allowed
sound.once('load', () => {
  sound.play()
})
```

Screenshot:

![image](https://github.com/tauri-apps/tauri/assets/15942946/7e2c6fdf-fcd3-45bb-8434-94b0aa72c344)

Error log from webview:

```
[Log] asset://localhost/%2Frun%2Fmedia%2Fthang%2FDATA2%2FMusic%2FTaylor%20Swift%2FTaylor%20Swift%20-%20Red%20(Deluxe%20Edition)%20(2018)%2F06.%20Taylor%20Swift%20-%2022.flac (player.js, line 16)
[Error] Origin http://localhost:1420 is not allowed by Access-Control-Allow-Origin. Status code: 200
[Error] Failed to load resource: Origin http://localhost:1420 is not allowed by Access-Control-Allow-Origin. Status code: 200 (/run/media/thang/DATA2/Music/Taylor Swift/Taylor Swift - Red (Deluxe Edition) (2018)/06. Taylor Swift - 22.flac, line 0)
[Error] XMLHttpRequest cannot load asset://localhost/%2Frun%2Fmedia%2Fthang%2FDATA2%2FMusic%2FTaylor%20Swift%2FTaylor%20Swift%20-%20Red%20(Deluxe%20Edition)%20(2018)%2F06.%20Taylor%20Swift%20-%2022.flac due to access control checks.
```

`npm run tauri info`:

```
> tauri info


[✔] Environment
    - OS: Arch Linux Unknown X64
    ✔ webkit2gtk-4.0: 2.40.1
    ✔ rsvg2: 2.56.0
    ✔ rustc: 1.69.0 (84c898d65 2023-04-16)
    ✔ Cargo: 1.69.0 (6e9a83356 2023-04-12)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-unknown-linux-gnu (default)
    - node: 18.16.0
    - npm: 9.5.1

[-] Packages
    - tauri [RUST]: 1.3.0
    - tauri-build [RUST]: 1.3.0
    - wry [RUST]: 0.24.3
    - tao [RUST]: 0.16.1
    - @tauri-apps/api [NPM]: 1.3.0
    - @tauri-apps/cli [NPM]: 1.3.1

[-] App
    - build-type: bundle
    - CSP: default-src 'self'; img-src 'self' asset: https://asset.localhost; media-src 'self' asset: https://asset.localhost
    - distDir: ../dist
    - devPath: http://localhost:1420/
    - framework: Vue.js
    - bundler: Vite
```

**Comment by wusyong** [comment]
webkitgtk still has issues to load video from custom URL scheme.
I also have a reproducible example in pure C.
https://github.com/wusyong/gtkbrowser/tree/video
But I couldn't track the upstream issue. I'll find some time to open one again (or find the one I commented before.)

**Comment by duydang2311** [comment]
I can load an image asset but not for audio/video asset even with correct configuration.

![image](https://github.com/tauri-apps/tauri/assets/34796192/5b6fa41f-def8-4e5b-8c73-a97f7c9f40e1)

```json
"protocol": {
        "all": false,
        "asset": true,
} // leave the assestScope here because I load audio file using tauri file picker

"security": {
      "csp": "default-src 'self'; media-src 'self' asset: https://asset.localhost"
},
```

**Comment by markfswe** [comment]
Dropped some info to https://bugs.webkit.org/show_bug.cgi?id=146351
Hope this will be fixed soon.

@lucasfernog is it possible to find which version was working for you? Even upstream bug was opened in 2015, maybe there where some changes that made it work for some versions. I am ready to compile previous versions of webkit, test it and go through changelog & git diff to help solving this faster

**Comment by lucasfernog** [comment]
Unfortunately I have no idea :( @ZerdoX-x

**Comment by foodornt** [comment]
I can affirm that I'm facing a similar issue to [duydang2311](https://github.com/tauri-apps/tauri/issues/3725#issuecomment-1626357908), I am able to load images through the asset protocol, although unable to load PDF documents using the pdfJS library
![image](https://github.com/tauri-apps/tauri/assets/40924367/d7784a6f-95fc-42bc-bcf6-09201793d443)

**Comment by xxidbr9** [comment]
@foodornt im using this setup to handle PDF files, and im using pdf js as well with local vendor files, this is work for all desktop OS (window, Mac, and Linux), add fs to allow the list and add the scope  

<img width="1093" alt="Screenshot 2023-08-20 at 06 44 07" src="https://github.com/tauri-apps/tauri/assets/51733515/fac15451-e424-417c-98ef-5963442d781b">
<img width="1680" alt="Screenshot 2023-08-20 at 06 52 17" src="https://github.com/tauri-apps/tauri/assets/51733515/c51f33c4-1bf4-44e2-9d27-7069c33cfa4b">

**Comment by foodornt** [comment]
@xxidbr9 it's very kind of you to stop by and share with us your setup! Thank you!
It appears that my issue is because I'm using react dev server (which is running on port 1420) and trying to access port 80 (Cross Origin) which is again being blocked by that dev server

![image](https://github.com/tauri-apps/tauri/assets/40924367/84769e1c-147e-4eda-ad6c-7455b7adee70)

Edit: there's no server running on port 80 anyways 🤔 I wonder what's serving files in the first place

**Comment by markfswe** [comment]
@foodornt looks like your problem is not related to this issue anyways.

**Comment by xxidbr9** [comment]
> @xxidbr9 it's very kind of you to stop by and share with us your setup! Thank you!
> 
> It appears that my issue is because I'm using react dev server (which is running on port 1420) and trying to access port 80 (Cross Origin) which is again being blocked by that dev server
> 
> 
> 
> ![image](https://github.com/tauri-apps/tauri/assets/40924367/84769e1c-147e-4eda-ad6c-7455b7adee70)
> 
> 
> 
> Edit: there's no server running on port 80 anyways 🤔 I wonder what's serving files in the first place
> 
> 
> 
> 

share your allowList, for the assets file it using asset protocol not http, so you need to add and add scope as well in fs 
```json
      "protocol": {
        "asset": true,
        "assetScope": ["**"]
      }
```

**Comment by foodornt** [comment]
@ZerdoX-x it could be, I'll open another one after I retry.
@xxidbr9 it's asset:// and not http:// request. I used the convertFileSrc function, and when debugging the URL, it indeed starts with asset://. The allowList is exactly like the one you provided.

**Comment by aaalloc** [comment]
Up,

Anyone have a workaround on this issue in Linux environment ?

**Comment by markfswe** [comment]
@Siirko yes. you can try to read the thread and find references from commits from other repos to this issue

**Comment by aaalloc** [comment]
For the people who are on Linux, images can be loaded through asset protocol (https://github.com/tauri-apps/tauri/discussions/1438) but not for videos and audio due to webkit lib issue (see https://bugs.webkit.org/show_bug.cgi?id=146351, last email). Someone on tauri discord found a workaround for audio file (this maybe applicable to other type of file), you just have to load the file with fs and create a Blob which you create an DataURL with it

Example : 

```js
  import { readBinaryFile } from "@tauri-apps/api/fs"
  const [urlAudio, setUrlAudio] = useState<string>("")
  return (
    <>
      <Button
        variant="ghost"
        onClick={() => {
          const filePath = .....
          void readBinaryFile(filePath)
            .catch((err) => {
              console.error(err)
            })
            .then((res) => {
              const fileBlob = new Blob([res as ArrayBuffer], { type: "audio/mpeg" })
              const reader = new FileReader()
              reader.readAsDataURL(fileBlob)
              const url = URL.createObjectURL(fileBlob)
              setUrlAudio(url)
            })
        }}
        className="w-full justify-start font-normal"
      >
        ok
      </Button>
      <audio controls>
        <source src={urlAudio} type="audio/mpeg" />
      </audio>
    </>
  )
  ```

You just need to setup tauri.conf.json permission for letting fs reading files with ``"scope": ["**"]``

EDIT : you could do even better with a simple fetch, you will have to add linux-protocol-headers to the features of tauri your cargo.toml file 
```
...
tauri = { version = "1.4", features = [...., "linux-protocol-headers"] }
....
```
And the code simplifies to 
```js
          const filePath = "/home/yanovskyy/Musique/Shinigami.m4a"
          fetch(convertFileSrc(filePath))
            .then((res) => {
              // create blob url from response
              res
                .blob()
                .then((blob) => {
                  const url = URL.createObjectURL(blob)
                  setUrlAudio(url)
                })
                .catch((err) => {
                  console.error(err)
                })
            })
            .catch((err) => {
              console.error(err)
            })
```
This, I think, is the best workaround for small video/audio file this moment because readBinaryFile slow down the process and there is memory problem. If you have big files (up to GB), well rip ...

**Comment by foodornt** [comment]
This is quite annoying
 I hope webkit lib solves it soon, although would still be problematic for users who haven't updated the lib

**Comment by amosjyng** [comment]
I'm not sure if this is the same bug, but I'm unable to play bundled audio files. I'm using SvelteKit and doing

```html
<audio controls>
  <source src="/switch.ogg" type="audio/ogg" />
</audio>
```

and placed the corresponding sound file in `src-svelte/static/switch.ogg`. I can play this sound on `yarn tauri dev` but not the release build. Other assets such as font files or images in the static directory load just fine.

If instead I set the audio source to be some random file from https://en.wikipedia.org/wiki/Category:Wikipedia_Ogg_files, then that external sound file plays successfully even on the release build.

If instead I place the file in `src-svelte/src/lib/sounds/switch.ogg` and try to programmatically play the sound:

```svelte
<script lang="ts">
  import clickSound from "$lib/sounds/switch.ogg";
  ...
  function playClick() {
    const audio = new Audio(clickSound);
    audio.play();
  }
</script>

<button
  ...
    on:click={playClick}
  >
    ...
</button>
```

I get the error

```
0:00:10.166424993 20275 0x559a88e1aa00 WARN            uridecodebin gsturidecodebin.c:1409:gen_source_element:<uridecodebin0> error: No URI handler implemented for "tauri".
```

where `console.log(clickSound)` gives me

```
tauri://localhost/_app/immutable/assets/switch.326f3a0f.ogg
```

`convertFileSrc` doesn't appear to help either.

**Comment by thewh1teagle** [comment]
Does it happens in Apple webkit as well?
https://bugs.webkit.org/show_bug.cgi?id=146351

Looks very related to the bug in Linux
The deeper issue: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3021

Looks like both `Apple` `webkit` and `webkitgtk` uses `gstreamer`
So it most likely an issue in `gstreamer`

Update:
It happens in `MacOS` `Webkit` as well!

![image](https://github.com/tauri-apps/tauri/assets/61390950/38e8ae31-7a08-4544-a637-2cb4a8305b6c)

**Comment by FabianLars** [comment]
gstreamer is only used on Linux. The issue in your case should be that you're providing a plain file path while it needs the asset: protocol (via convertFileSrc).

**Comment by thewh1teagle** [comment]
@FabianLars 
Looks like apple's webkit does uses `gstreamer`.

[Source/WebCore/platform/mediastream/gstreamer](https://github.com/WebKit/WebKit/tree/9d8dc32fa51b3809870a01954ac019de3f6de688/Source/WebCore/platform/mediastream/gstreamer)

**Comment by FabianLars** [comment]
That codebase is shared between apple, webkitgtk, and wpe (and more?). So having a gstreamer folder doesn't mean much for the macos port, especially if we take a look at what other folders are in that mediastream folder.

**Comment by thomaskvnze** [comment]
Guess I stumbled into the same issue. 

I have build an app with tauri and sveltekit for the raspberry pi. Everything works so far except playing an embedded mp4 video in the bundled app for the pi.

The video works on all other platforms in their respective bundled version. On the raspi, I see the play button crossed out. There is no error in the console. I can see that the source section in the inspector contains the video. The src link uses the tauri protocol in the bundled version. When I change the src url in the inspector dom to the same video hosted on a server, it plays the video. I assume this is the same custom schema protocol issue. 
![image](https://github.com/tauri-apps/tauri/assets/9097796/aac074f8-2b2e-4dc0-9039-a6d378bdcd8a)

**Comment by superkelvint** [comment]
I'm running into this issue on Linux (MP3 audio playback works with dev but not release build), except its happening not just for assets but also remote MP3 URL playback. 

Does anyone know of a workaround on Linux?

**Comment by niraj-khatiwada** [comment]
Any updates on this? I've been holding off a feature in Linux that allows video play from local asset. Can anyone from the team please respond if there are any alternatives?

**Comment by FabianLars** [comment]
Still the same alternatives/workarounds as mentioned earlier in this issue:
1. Load small files via fs functions as binary
2. Large files require a http server on the rust side (or a sidecar in your language of choice i guess) so that the file will be loaded over http instead.

Because we do not want to add a http server into tauri itself (assuming nobody changed their minds), we're stuck with this until we have an alternative for webkitgtk.

**Comment by markfswe** [comment]
> Does anyone know of a workaround on Linux?

@Siirko @superkelvint please take a look at references to this issue, they are available in the discussion above. Please do not bump same issue with same questions.

https://github.com/luxuereal/Alexandria_EBook_Tauri/commit/1050664fafc5d0b720dda7fa112d27cd789c8dbf
https://github.com/tranxuanthang/lrcget/commit/0a2fe9943e40503a1dc5d9bf291314f31ea66941
https://github.com/vjousse/pomodorolm/commit/61e90557314d1ad675ee858c33f1a3d17a8aa93b

this might be helpful.

---

@niraj-khatiwada 
> Any updates on this?

https://bugs.webkit.org/show_bug.cgi?id=146351 Here is the original bug report and discussion in webkit bug tracker. Tauri didn't do anything on it's side yet, and I don't think it can do for now. As it will break one of the core feature: working without running local server.

**Comment by ZerdoX-x** [comment]
> Any updates on this?

https://bugs.webkit.org/show_bug.cgi?id=146351 Here is the original bug report for webkit.

**Comment by niraj-khatiwada** [comment]
> > Does anyone know of a workaround on Linux?
> 
> @Siirko @superkelvint please take a look at references to this issue, they are available in the discussion above. Please do not bump same issue with same questions.
> 
> [luxuereal/Alexandria_EBook_Tauri@1050664](https://github.com/luxuereal/Alexandria_EBook_Tauri/commit/1050664fafc5d0b720dda7fa112d27cd789c8dbf) [tranxuanthang/lrcget@0a2fe99](https://github.com/tranxuanthang/lrcget/commit/0a2fe9943e40503a1dc5d9bf291314f31ea66941) [vjousse/pomodorolm@61e9055](https://github.com/vjousse/pomodorolm/commit/61e90557314d1ad675ee858c33f1a3d17a8aa93b)
> 
> this might be helpful.
> 
> @niraj-khatiwada
> 
> > Any updates on this?
> 
> https://bugs.webkit.org/show_bug.cgi?id=146351 Here is the original bug report and discussion in webkit bug tracker. Tauri didn't do anything on it's side yet, and I don't think it can do for now. As it will break one of the core feature: working without running local server.

Thanks for the information. I didn't realize it was the issue on webgtk gstreamer itself. I understand that Tauri cannot do anything about it. Looks like it's already been 9 years since the issue was raised and I don't think it is going to be fixed anytime soon. So I'll just spin up a http server as a workaround.

**Comment by GiorgioBaldelli** [comment]
A quick workaround that I resorted to for audio files:

``` 
let sound = new Audio("data:audio/mpeg;base64,<insert base64 converted audio string>");
sound.currentTime = 0;
sound.play();
```

Less than ideal, but does the job for small/short audio files.

**Comment by ImUrX** [comment]
cursed way to fix this issue:
```js
const blobUrl = URL.createObjectURL(await fetch(asset).then(res => res.blob()))
const sound = new Audio(blobUrl)
```

**Comment by loziniak** [comment]
Perhaps it could be worked around by Tauri by allowing to register custom response generator under a path of default scheme? Something like `tauri::Builder::default().register_uri_path("/videos", |_app, request| { ...` , where all requests going to `/videos` would be handled by custom code?

Could [WebviewBuilder::on_web_resource_request](https://docs.rs/tauri/2.0.0-rc.16/tauri/webview/struct.WebviewBuilder.html#method.on_web_resource_request) be the solution?

**Comment by FabianLars** [comment]
This is exactly how the default tauri and asset protocols work and the root cause of the issue here so yet another protocol with the same implementation would not help anyone.

**Comment by loziniak** [comment]
I understand. I assumed this is possible: https://github.com/tauri-apps/wry/issues/1087
And that in release mode files are not loaded all by custom protocol, like `tauri:` .

**Comment by FabianLars** [comment]
Yeah that would be cool but only windows supports modifying arbitrary http requests/responses like that (and it's also how the current custom protocol implementation works on windows). The current macOS and Linux webviews only support monitoring requests.

**Comment by eleroy** [comment]
Hello I can confirm that it works well @ImUrX solution for video. My project is hybrid react-native+expo+tauri and I use a method like this:
```typescript
const videoFromURL = async (url: string) => {
    if (Platform.OS == "web") {
      const blobUrl = URL.createObjectURL(
        await fetch(url).then((res) => res.blob())
      );
      return blobUrl;
    }
    return url;
  };
```

To load video asset into the video player. (My problem was to load bundled asset into the dist folder)

**Comment by ImUrX** [comment]
Be careful with the generated blob URL, it causes a memory leak unless you free it again later manually!
More info in https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static

**Comment by haexhub** [comment]
Is there any news on this? 
Loading the whole (large) audio/video file as blob isn't really a solution.
In the [tauri docs](https://github.com/tauri-apps/tauri/blob/dev/packages/api/src/core.ts#L257)  is an example, how to load a video file via asset protocol. I tried this for audio without success.

**Comment by FabianLars** [comment]
> In the [tauri docs](https://github.com/tauri-apps/tauri/blob/dev/packages/api/src/core.ts?rgh-link-date=2025-05-21T18%3A15%3A20.000Z#L257) is an example, how to load a video file via asset protocol. I tried this for audio without success.

Yeah it doesn't because that's exactly what this issue is about. And no, there were no updates and there very likely won't be any until we drop webkitgtk.

I would have liked to see something similar to our existing localhost plugin that serves videos over http but that never happened and at this point probably never will.

**Comment by Bandonker** [comment]
I've encountered this exact issue while building audio applications in Tauri. The `asset://` protocol limitations on Linux (webkit2gtk) are particularly frustrating so I ended up creating a library specifically to solve this.

 NPM:  [desktop-audio-proxy](https://www.npmjs.com/package/desktop-audio-proxy)
 github: [DAP](https://github.com/Bandonker/desktop-audio-proxy)

  As mentioned in this issue, Tauri's `asset://` protocol has webkit2gtk limitations that prevent reliable audio/video streaming, especially on Linux.
   The CSP and protocol restrictions make it unnecessarily complex to handle media files.

  How it's used:

  ```typescript
  import { TauriAudioService } from 'desktop-audio-proxy/browser';

  // Automatic setup - handles convertFileSrc internally
  const audioService = new TauriAudioService({
    autoDetect: true
  });

  // Works with any local audio/video file
  const streamUrl = await audioService.getStreamableUrl('/path/to/local/audio.mp3');
  audioElement.src = streamUrl; // Now plays reliably across all platforms

  ```

How to Use DAP with Videos:

  ```typescript

  // Works with video URLs just like audio
  const videoUrl = await client.getPlayableUrl('https://cors-restricted-site.com/video.mp4');
  videoElement.src = videoUrl; // Now plays without CORS issues

  Tauri Video Example:

  import { TauriAudioService } from 'desktop-audio-proxy/browser';

  const service = new TauriAudioService();

  // Local video file
  const localVideo = await service.getStreamableUrl('/path/to/local/video.mp4');
  videoElement.src = localVideo;

  // Remote video with CORS issues  
  const remoteVideo = await service.getStreamableUrl('https://external-site.com/video.webm');
  videoElement.src = remoteVideo;
  ```

  Instead of fighting with protocol limitations, the library serves media files through a local HTTP proxy that webkit2gtk handles perfectly. It's the same approach that works reliably in browsers, adapted for desktop applications. This has solved the asset loading issue for multiple of my Tauri projects dealing with audio/video streaming across different platforms

**Comment by aaalloc** [comment]
After some times, I've found a solution with Gstreamer

https://github.com/user-attachments/assets/6f0d7a0a-8520-46e4-90d3-0aa8c20ec653

Here a video loading through asset://, it also load audio file (not shown in this video)


I'm not certainly sure if everything works with my solution but it seems to load videos ! I think that solution would finally close this really old issue.

I'm available to discuss about the solution on discord with maintainers

**Comment by niraj-khatiwada** [comment]
> After some times, I might have found a solution related to Gstreamer
> 
>  Capture.video.du.2025-10-28.21-22-57.mp4 
> I'm not certainly sure if everything works with my solution but it seems to load videos ! I'm available to discuss about the solution on discord and I think that solution would finally close this really old issue.

Is it something we can pass as a flag when running the app on runtime or it's something that should be included in Tauri internal code?

**Comment by aaalloc** [comment]
> > After some times, I might have found a solution related to Gstreamer
> > 
> >  Capture.video.du.2025-10-28.21-22-57.mp4 
> > I'm not certainly sure if everything works with my solution but it seems to load videos ! I'm available to discuss about the solution on discord and I think that solution would finally close this really old issue.
> 
> Is it something we can pass as a flag when running the app on runtime or it's something that should be included in Tauri internal code?

To fix this at this moment, it could be included in Tauri internal code which I made a PR for it. Otherwise another solution would be to check with WebKit maintainers on how this could be generalised and used internally. But for that, it may take a while...

**Comment by kartonrad** [comment]
Heads up:
The bug is in the Chromium Webview itself.

Check out this chromium issue:
https://issues.chromium.org/issues/40739128

I walked through the code and i believe that i may have found the issue that the chromium people were talking about.

Basically, audio and video players make 'Range' request, but the shouldInterceptRequest Function used for implementing the Custom Protocol Handler, assumes that your asset server returns the entire video at once.
It ignores the Content-Range response header and seeks to the start prosition of the Range request header.

It basically assumes that you aren't really giving it a response body, it assumes you're just passing in the InputStream to a file, nothing more.

It seems like this will never work without a fix in WebView.

In the meantime, you HAVE to use localhost, or implement a custom video/audio player, so that you can intercept the request -
And rename the Range header to something like X-Range, so your backend can handle it.

Sucks.
Related issue:
https://github.com/issues/recent?issue=tauri-apps%7Ctauri%7C12019

Related Discussion:
https://github.com/orgs/tauri-apps/discussions/12243

My comment on the chromium thread:
https://issues.chromium.org/issues/40739128#comment30

**Comment by aaalloc** [comment]
> Heads up:
> The bug is in the Chromium Webview itself.
> 
> Check out this chromium issue:
> https://issues.chromium.org/issues/40739128
> 
> I walked through the code and i believe that i may have found the issue that the chromium people were talking about.
> 
> Basically, audio and video players make 'Range' request, but the shouldInterceptRequest Function used for implementing the Custom Protocol Handler, assumes that your asset server returns the entire video at once.
> It ignores the Content-Range response header and seeks to the start prosition of the Range request header.
> 
> It basically assumes that you aren't really giving it a response body, it assumes you're just passing in the InputStream to a file, nothing more.
> 
> It seems like this will never work without a fix in WebView.
> 
> In the meantime, you HAVE to use localhost, or implement a custom video/audio player, so that you can intercept the request -
> And rename the Range header to something like X-Range, so your backend can handle it.
> 
> Sucks.
> Related issue:
> https://github.com/issues/recent?issue=tauri-apps%7Ctauri%7C12019
> 
> Related Discussion:
> https://github.com/orgs/tauri-apps/discussions/12243
> 
> My comment on the chromium thread:
> https://issues.chromium.org/issues/40739128#comment30
> 
> 

I guess another issue has to been opened, we are talking about WebKitGTK for Linux in this issue, not android case with Chromium webview

**Comment by kartonrad** [comment]
oops! what am i talking about, the platform is right there in the sidebar!

**Comment by itsmeadarsh2008** [comment]
Is this issue still active?

**Comment by aaalloc** [comment]
> Is this issue still active?

@itsmeadarsh2008 

Yes, but #14402 solves the issue. It has been reviewed by some people but it needs to be approved by maintainers

**Comment by itsmeadarsh2008** [comment]
> > Is this issue still active?
> 
> [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> 
> Yes, but [#14402](https://github.com/tauri-apps/tauri/pull/14402) solves the issue. It has been reviewed by some people but it needs to be approved by maintainers

Will this fix for streaming audio, not just local?

**Comment by rzru** [comment]
hey folks, is there any progress on this one?

**Comment by itsmeadarsh2008** [comment]
> hey folks, is there any progress on this one?

bro, I am in the middle of nowhere for my music app. Finding this out after switching from Windows to Linux 😢 I wish Tauri had switchable web engines like electrobun does.

**Comment by aaalloc** [comment]
> > > Is this issue still active?
> > 
> > 
> > [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> > Yes, but [#14402](https://github.com/tauri-apps/tauri/pull/14402) solves the issue. It has been reviewed by some people but it needs to be approved by maintainers
> 
> Will this fix for streaming audio, not just local?
@itsmeadarsh2008 

What do you mean by that ?

**Comment by itsmeadarsh2008** [comment]
> > > > Is this issue still active?
> > > 
> > > 
> > > [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> > > Yes, but [#14402](https://github.com/tauri-apps/tauri/pull/14402) solves the issue. It has been reviewed by some people but it needs to be approved by maintainers
> > 
> > 
> > Will this fix for streaming audio, not just local?
> > [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> 
> What do you mean by that ?

I mean, loading the audio from the internet, not just local. See, I have tried AppImage it's working but app is super laggy for some reason, current issue is not working with rpm and deb builds.

**Comment by aaalloc** [comment]
> > > > > Is this issue still active?
> > > > 
> > > > 
> > > > [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> > > > Yes, but [#14402](https://github.com/tauri-apps/tauri/pull/14402) solves the issue. It has been reviewed by some people but it needs to be approved by maintainers
> > > 
> > > 
> > > Will this fix for streaming audio, not just local?
> > > [@itsmeadarsh2008](https://github.com/itsmeadarsh2008)
> > 
> > 
> > What do you mean by that ?
> 
> I mean, loading the audio from the internet, not just local. See, I have tried AppImage it's working but super laggy for some reason, current issue is not working with rpm and deb builds.

Then it is a different issue, this issue and the fix is related to load video/music files from `asset://`

**Comment by markfswe** [comment]
> Then it is a different issue, this issue and the fix is related to load video/music files from `asset://`

3 years in the thread.

This is bug [146351](https://bugs.webkit.org/show_bug.cgi?id=146351) on bugs.**webkit**.org
It was reported back in 2015 and currently has 10 on CC list.

That is if we suppose github issue title says "does not work, on Linux", as it should. 

Afaik the easiest workaround currently is - to open a localhost port, if user is on Linux, and serve it media/audio/assets. There were some examples of workaround, see commits and PR's referencing this issue directly in the feed.

though @aaalloc above might share some more details

**Comment by aaalloc** [comment]
@markfswe 
> This is bug [146351](https://bugs.webkit.org/show_bug.cgi?id=146351) on bugs.webkit.org
It was reported back in 2015 and currently has 10 on CC list.

Technically, this is not a bug from WebKit nor Gstreamer (which is used for audio/video processing), it's just that something was missed when `asset://` was implemented 

> Afaik the easiest workaround currently is - to open a localhost port, if user is on Linux, and serve it media/audio/assets. There were some examples of workaround, see commits and PR's referencing this issue directly in the feed.

There is a even more easier fix for linux without rewriting code logic, see #14402 for more details but the fix is just a shared object file at the right path and that's it

**Comment by markfswe** [comment]
Crazy work, nice to see.

> see [#14402](https://github.com/tauri-apps/tauri/pull/14402) for more details but the fix is just a shared object file at the right path and that's it

Appreciate for the update

**Comment by Leinonen96** [comment]
I can also confirm this is still reproducible on current Tauri v2.

### Environment

- Fedora 44
- KDE Plasma 6.7.3 (reproduced on both Wayland and X11)
- WebKitGTK 2.52.5
- Tauri 2.11.5

During debugging I observed:

- `fetch(convertFileSrc(...))` succeeds (HTTP 200) and returns the complete file.
- `video.canPlayType(...)` reports the tested MP4 codec is supported.
- `video.src = convertFileSrc(...)` fails with `MEDIA_ERR_SRC_NOT_SUPPORTED (4)`.
- Using a Blob URL changes the failure to `MEDIA_ERR_DECODE (3)`, rendering the first frame before playback aborts.
- Serving the exact same local file through a localhost HTTP server running inside the Tauri backend immediately restores normal playback and seeking.

I independently arrived at the same workaround (serving the media over localhost instead of `asset://`), which solved the issue for my application.

**Comment by blurymind** [comment]
serving the files over localhost is half broken on tauri. I found that video streaming cuts off half way no matter what i tried

**Comment by up9t** [comment]
Doesn't work in Fedora 44 + Gnome. Tauri V2.

Reproducible code:

```typescript
import audioSrc from "../assets/sounds/abc.mp3";

function playAudio() {
  new Audio(audioSrc).play();
}
```
